doc: require two approvals to land changes

Trott opened the pull request:
First in a series that will hopefully streamline/simplify our policies. I don't want this to slip in unnoticed and it's a significant procedural change being proposed, so: @nodejs/collaborators

Currently, changes require approval by one Collaborator in most cases.
However there are situations where two approvals are required. For
example, breaking changes require two approvals from TSC members. And
fast-tracking a request requires two approvals.

Additionally, although only one approval is strictly required, in
practice, we nearly always seek a second approval when there is only
one.

Lastly, concerns have been raised about (perhaps unintentionally) gaming
the one-approval system by suggesting a change to someone else, and then
approving that change when the user submits a pull request. This
resolves (or at least mitigates) that concern.

Refs: https://github.com/nodejs/node/issues/19564

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

ChALkeR commented:
LGTM, though I am curious how many (and which exact) changes landed with only one collaborator review this year. That might give us some hints what changes might become harder to land.

jasnell commented:
I'm generally -1 on this as I don't see the change as being necessary. 1 approval should be enough if the change is relatively constrained and the list of people who have the necessary domain knowledge to adequately review is small.

Trott commented:
@ChALkeR Here are all the ones since the start of July.

d3d54aa8ef6e4df04507f3d3a1b891c863eb497b
284caaa852194f7418d24c44f2df58614437f3cc
3095eecc4748da4ce7ac70e2b352ddba6c4c4deb
87f767160d103114eff154e98bde69caa178c59f
f7454a53f1318837ea2ca8f2ad495a9869b80e81
a706456f71e28d4317beb0bf9da7a31adbf11ca5
ededb4b510d37fa73582f71d7bd6d3a5bb046be2
57e301539bff2599974b7269a56377330c9b730e
85b0f1649d07294afea6cfbb3082c18a3dc2079a
f386c7e41c955a175fe338e697a4df5921b4cfad
e021ca28ea6dc87094e97afd3aec6fd8cecd571d

That's 11 commits. For perspective, there were 352 commits in that time, so were talking about approximately 0.3% of commits.

bengl commented on COLLABORATOR_GUIDE.md:
```diff
@@ -111,10 +111,9 @@ comment that explains why the PR does not require a CI run.
-sufficient expertise who is able to take full responsibility for the
-change. In the case of pull requests proposed by an existing
-Collaborator, an additional Collaborator is required for sign-off.
+All pull requests must be reviewed and accepted by two Collaborators with
```
This may be a good opportunity to clarify that we mean *at least* two, and not exactly two.

joyeecheung commented:
Maybe we can make exceptions for certain subsystems? AFAICT crypto may need that.

Trott commented:
> 1 approval should be enough if the change is relatively constrained and the list of people who have the necessary domain knowledge to adequately review is small.

@jasnell [The last time we reviewed changes that had only one approval](https://github.com/nodejs/node/issues/19564#issuecomment-379392324), none of them fit that description. I think preserving the one-approval rule for that reason is preserving it for a problem that we don't actually have. As I wrote at that time:

> I would conclude that we're not using group @-mentions enough to draw attention to PRs. Seems to me like the proposed rule change would encourage people to surface PRs to relevant groups more often.

Trott commented:
> Considering those numbers and the types of commits, I do not see this as needed.

@mcollina With only one exception (85b0f1649d07294afea6cfbb3082c18a3dc2079a is), all of the commits seem like they could have obtained more reviews with just a little effort. I mean, look at f386c7e41c955a175fe338e697a4df5921b4cfad.

Time to repeat [the exercise from last time](https://github.com/nodejs/node/issues/19564#issuecomment-379392324):

*  d3d54aa is a straightforward Makefile change that could have gotten a second review. Certainly the issue here isn't that there aren't people with sufficient expertise to review it. Indeed, I'd be surprised if that PR was unable to get four or five additional reviews.

* a706456f71e28d4317beb0bf9da7a31adbf11ca5, a706456f71e28d4317beb0bf9da7a31adbf11ca5 and 284caaa852194f7418d24c44f2df58614437f3cc are actually all the same commit. It's a V8 backport from upstream that got re-applied when we updated V8. Anyway, we routinely get multiple reviews on V8 backports and there is little doubt that additional reviews could have been amassed for that were it required. The relevant PR was also open for less than the required 72 hours on a weekend, so that may be the reason for only one review.

* 87f767160d103114eff154e98bde69caa178c59f and 386c7e41c955a175fe338e697a4df5921b4cfad are simple doc change. It would have been simple to get more reviews for that.

* 3095eecc4748da4ce7ac70e2b352ddba6c4c4deb is not complex and another review would have certainly been attained had it been sought. No groups were @-mentioned and @ChALkeR even chimed in with an after-the-fact LGTM.

* f7454a53f1318837ea2ca8f2ad495a9869b80e81 is the one where things may have been delayed. However, that one doesn't exactly make the case because it was landed in violation of the 48/72 hour rule. (It was open for less than 72 hours on a weekend.) It's entirely possible (likely?) that it would have obtained a second review had it been open long enough.

* With ededb4b510d37fa73582f71d7bd6d3a5bb046be2z and 57e301539bff2599974b7269a56377330c9b730e, there were no attempts to @-mention any relevant teams. I'd argue that indicates insufficient effort to get multiple reviews (which is OK in the current setup since that's all that's required--but doesn't exactly make it a foregone conclusion that additional reviews couldn't be obtained if they were sought).

* 85b0f1649d07294afea6cfbb3082c18a3dc2079a is a case where @tniessen tried repeatedly to get the attention of @nodejs/crypto but only got one review. So that's one commit that might have run into trouble with this process. 

* e021ca28ea6dc87094e97afd3aec6fd8cecd571d has no team ping. Seems like something that would have gotten more reviews if they were sought.

Trott commented:
> Maybe we can make exceptions for certain subsystems? AFAICT crypto may need that.

@joyeecheung Maybe, although that undermines the simplicity/consistency aspect. The only recent commit where reviews were actively sought and not obtained was a crypto one (85b0f16). On the other hand, that is the only crypto PR that has landed this year with only one review. And there have been a good number of crypto commits. So maybe it's not such a problem for crypto after all? Kind of interested in what @tniessen thinks, as he's the one that was trying to get reviews and only got one.

refack commented:
IMHO velocity on it's own is not a goal. Having more review should be a great tool to improve code quality, which should be of higher priority.

I'd also like for us to put more emphasis on the other parts of that paragraph:
> All pull requests must be reviewed and accepted by a Collaborator with __sufficient expertise__ who is able to __take full responsibility__ for the change.

It seems to me that this will go to a vote. In that case we might want to consider expanding the scope of this change (longer review window, cap maximum of approvals, require /CC of "code owners")

addaleax commented:
> Having more review should be a great tool to improve code quality, which should be of higher priority.

This. :100: If we can’t get enough reviewers, then we really should force ourselves put effort into making more people familiar with the relevant code.

jasnell commented:
Having more reviewers per PR is a fantastic goal. Imposing additional process does not accomplish that, encouraging more folks to get involved does.

Trott commented:
> Having more reviewers per PR is a fantastic goal. Imposing additional process does not accomplish that

* This does not impose additional process. Increases existing requirements from one review to two reviews? Yes. Imposes additional process? No. It clarifies and minimally increases requirements to what is in fact the _de facto_ for approximately 97% of our commits.

* It is self-evident and beyond any doubt that this rule will absolutely increase the number of reviewers for PRs. It is impossible IMO to look at https://github.com/nodejs/node/pull/21670 and reasonably believe that this rule wouldn't have increased the number of reviews that PR got at least from one to two. You could argue that it is not worthwhile to get more reviews on such a PR--I would dispute that, but one can reasonably argue it--but you'd have to embrace some pretty unlikely scenarios to conclude that PR would not get a second review. Ping the documentation group and/or the http group and you'd have a second review in no time. (Or perhaps even more likely, addaleax would have reviewed it before landing it.)

benjamingr commented:
One problem is that for some PRs in less glorious and well known parts of the code getting even one review is hard.

Can we get an estimate of how many pull requests did not get a single review in 3 days?

benjamingr commented:
Another issue is that people might not be as responsible when reviewing (e.g run the actual code and play with the changes) when reviewing if they think at least one other person is responsible. I think this has a name in psychology.

Trott commented:
> but I eventually would like us to work towards a better review process.

@benjamingr I'm interested in any concrete thoughts you or anyone else has about what that might look like. Happy to talk offline with anyone about that. (Probably best to keep this PR focused on just the one narrow proposed change.)

vsemozhetbyt commented:
Can we consider one aspect here: how this new restriction and "with sufficient expertise who are able to take full responsibility for the change" part relates to rubber stump LGTM? I see RSLGTM periodically, sometimes we have only RSLGTMs in a PR. After this change, RSLGTMs number can be increased. Will we consider them as a sufficient approving?

tniessen commented:
> So maybe it's not such a problem for crypto after all? Kind of interested in what @tniessen thinks, as he's the one that was trying to get reviews and only got one.

85b0f1649d07294afea6cfbb3082c18a3dc2079a was landed with only one approving review, but it was also reviewed by @addaleax and @bnoordhuis. I think if we fail to get two approvals on a PR despite having 118 collaborators, that indicates deeper problems in our review process. One might be that the @nodejs/crypto team has about five or six active members, which is a horrendously small number compared to far more than a hundred collaborators.
In my experience, getting the required number of TSC approvals on semver-major crypto PRs has been far more challenging than getting a few collaborator reviews. (It took *weeks* to get TSC approvals on https://github.com/nodejs/node/pull/18138.)

BridgeAR commented:
Just as a note: I struggled multiple times to get reviews and our open PRs are growing all the time at the moment. However, this is not really impacted by having a single or two reviews but more often by getting reviews at all or not getting any reviews (and also by getting a LG after a change).

mcollina commented:
@Trott would having more reviews reduced the number of bugs, or flakyness of the tests? Has there been a revert because of those commits? 

I share what @tniessen thinks for crypto, and this reflect my experience on streams. In the past I had a lot of trouble to get collaborators to review some of the stream changes, and not because of lack of effort in pinging people.

I can live with some added text that provides an escape hatch: "if after 7 days a PR did not get 2 reviews, it is possible to land it with just one (this does not applies to semver-major changes)".

benjamingr commented:
I'd like to echo Ruben's sentiment - as someone who reviews a lot more than I commit - I think we should encourage more people to review and review more carefully.

BridgeAR commented:
I like the suggestion by @mcollina with the escape hatch.

Trott commented:
> -1 from my side because this may prove problematic for subsystems such as @nodejs/crypto where only a couple of people take ownership of the entire module and sometimes it's nearly impossible to get two legitimate reviews.

@ryzokuken There's been only one crypto commit all year that failed to get more than one review.

The perception of the "It's so hard to get more than one review for crypto (or streams or whatever)" problem is much greater than the actual occurrence.

That said, it *does* happen. Just not very often.

Trott commented:
Also: I can get behind the escape-hatch proposal. I'd favor something like N hours after first pinging a team or even specifically `@nodejs/collaborators` rather than N hours after the PR is opened but that's all details.

jasnell commented:
We already have a 48-72 hour mandatory wait period. Again, this is attempting to solve for something is not and has not been a problem.

Trott commented:
> Again, this is attempting to solve for something is not and has not been a problem.

This addresses multiple real problems. (It "solves" none of them probably, but mitigates most or all of them.)

Summarizing (and sometimes expanding on) points made by different folks above:

* Some Collaborators review much more carefully than others. Requiring two reviews could mitigate the impact of what might be perceived by these Collaborators as rubber-stamp reviews that are given too liberally.
* Things land without sufficient effort to ping teams to get informed expert/stakeholder opinions.
* The problem described in https://github.com/nodejs/node/issues/19564.
* Too many variants in our rules makes it easy for even our most active contributors to forget or confuse rules, such as forgetting that there need to be two (not one) Collaborator approvals for fast-tracking a pull request. This helps at the margins by eliminating the "is it two or is it one" question. It's always two. (By itself, perhaps not a big win. Combined with other adjustments, we may get somewhere. Speaking of which...)

> We already have a 48-72 hour mandatory wait period.

That's another rule that is violated by active contributors and needs to be fixed. I was planning on staggering these PRs, but I'm thinking I should open them faster to see which ones get traction and which ones don't.

Trott commented:
> @Trott Has there been a revert because of those commits?

@mcollina Would it change your mind about this if the answer was "yes"?

Trott commented:
> @mcollina Would it change your mind about this if the answer was "yes"? [EDIT: Honest question. Totally understandable if the answer is "probably not"! I asked because I don't want to dig through logs if it's not going to change anyone's mind anyway.]

Turns out digging for it was easy so I did it.

Last year, there were four commits reverted that had only one `Reviewed-By`. However all four were from the same pull request (https://github.com/nodejs/node/pull/17407).

I imagine that one pull request is not enough to move anyone from the "no" camp to the "yes" camp on this proposal.

mcollina commented:
@Trott thanks for digging up the data. I do not think this PR is much needed, but I'll be ok with it if it has an escape hatch after the ping of the relevant team (as listed in https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#who-to-cc-in-the-issue-tracker).

> Things land without sufficient effort to ping teams to get expert/stakeholder opinions.

This is the problem we should solve. I do not think that increasing the time/requiring two signoffs are good solutions. Requiring the team/experts on that subsystem is. In our policies all LGTMs are born equal, however different expertise in specific areas carry different weight.

jasnell commented:
Rather than continuing to debate it, I've proposed what I believe is a better approach: https://github.com/nodejs/node/pull/22282

mscdex commented on COLLABORATOR_GUIDE.md:
```diff
@@ -111,10 +111,10 @@ comment that explains why the PR does not require a CI run.
-Collaborator, an additional Collaborator is required for sign-off.
+All pull requests must be reviewed and accepted by at least two Collaborators
+with sufficient expertise who are able to take full responsibility for the
+change. Approval must be from Collaborators who are not authors of the proposed
```
Perhaps "authors" is a bit vague here? Does that mean the individual who opened the PR, anyone that contributed a commit to the PR, anyone that requested the changes that led to the opening of the PR, or something else (including a combination of the above)?

gireeshpunathil commented:
- Is there an evidence on the existing rules / process caused harm to our project? (and in ways that broke a user)
 - As a collaborator, I trust other collaborators, for us to do the right thing; does this change put us on a bad light?
 - I review PRs on a premise that I am fully responsible for the changes and feel the ownership - irrespective of existing and subsequent reviews. Will this change distort / split that ownership?
 - `gaming the system`: Unfortunately I am unable to follow this: Is there an example available?
 - on the other hand, if rules were broken in the past, how these new ones are immune from being broken?

I acknowledge that more reviews defenitely strengthen our code base, but stringent rules around review process may not help to improve the quality, but can potentially cause over-emphasis on process and can reduce the empowerment on collaborators. I feel more comfortable if these were proposed as best practices rather than rules.

mscdex commented:
> gaming the system: Unfortunately I am unable to follow this: Is there an example available?

@gireeshpunathil See the referenced link from the commit for more information on this one

jasnell commented:
> As a collaborator, I trust other collaborators, for us to do the right thing; does this change put us on a bad light?

This is the crux of the issue for me. Because of the history of the project, our processes and rules are fundamentally designed around the concept that collaborators do not trust one another. Our rules are designed to stop and delay and check. This PR is based on the premise that we should trust collaborators even less. I have a major problem with that.

I believe the effort that's going in to discussions around changing the rules is far better spent on further improvement of the automation tools that *actually do* make things easier for all collaborators. We should be finding ways of reducing friction and showing that we actually do trust our collaborators.

For instance, in the PR that discusses extending the minimum landing time for PRs to 72 hours, it was pointed out that a couple of PRs landed this weekend in less than 72 hours. The argument is made that it's somehow a problem that those didn't wait the full time. Personally, I couldn't care less that they landed inside the 72 hour mark because I do trust my fellow collaborators to do the right thing. None of the PRs that landed are controversial or had *any* other reason not to be landed outside of that one rule.

This is the heart of why I argue that this change is not actually necessary because I don't see it solving an actual problem and it just continues a culture of mistrust.

gabrielschulhof commented:
I'm not sure how https://github.com/nodejs/node/pull/21611 fits into this picture or how this modification will help it land. It came in about a month-and-a-half ago, and was commented on right away. Then it seems to have stalled for 20 days, at which point I noticed it and found it highly relevant to the multithreaded use case, so I made some review comments. An issue was reported that would be fixed by the modification the day after my comments. A week later, and after some discussion, I decided to push to the branch in order to address my earlier review comments, since the original author hadn't responded – which is OK. Then, over the course of the last week, I pinged increasingly wider circles of collaborators for a second review, and approved it myself. I still haven't received a second review.

Now, at this point, aside from the obvious they-have-no-time-for-this thought, I'm thinking: "Am I unable to convince other collaborators to review because the change is not important?" or "Is this one way of expressing that this commit is not to land?"

The reason I mention all this is because from this thread my general feeling is that getting more than one timely review seems not to be a challenge for anybody, whereas in this case it seems to me very much like a challenge.

Then again, I'm not famous for my patience 🙂

Trott commented:
@gabrielschulhof With a relatively small C++ PR like that one, you might have some luck pinging some of the usual suspects in IRC or email. You shouldn't *have* to do that, and I think pinging Collaborators first may be the right thing (which you did yesterday).  But if *that* doesn't get another review in the next 24 or maybe 48 hours, it would be a good candidate for "this is an example of a PR that will be held up by this change". (I would argue that it's a good PR to be held up by this change because the changes aren't so obscure that there can't be a couple of reviews.)

rubys commented:
I too would like to avoid the "lack of trust" nomenclature as it unnecessarily polarizes the discussion.

Similarly, I don't think that the real question is one of whether there should be more or less reviews, but as to the timing of the reviews.

To draw an analogy, when I started in open source, CVS was the predominant VC software.  In order to contribute, you needed to be added as a committer, and you locked files preventing others from making changes while you did your work.  In short, the tool dictated the workflow.

Over time, this was replaced by subversion.  You still had to be added as a committer, but locking no longer was the common way to go, instead you resolved conflicts when you were ready to commit.

Over time, git has displaced a lot of svn usage.  You no longer need to be added as a committer in order to contribute.  You still are free to impose whatever workflow you wanted, but at this point the tool is no longer in the way.

Reviews are definitely a good thing, but I routinely see pull requests in node.js go multiple weeks, and often see the need to rebase the changes (sometimes multiple times) due to the delay.  Even with reviews, I have personally already introduced a regression, and once I was made aware of it, I started work on another pull request to address it.  Also, I will add that in most cases where people are making review comments, I would see them instead directly making the change that they desire in other projects.  

Just a different perspective, from someone who has worked on numerous other projects.

gabrielschulhof commented:
@rubys I'm sure there are times where a collaborator could simply push to the PR branch for no-brainer changes, and perhaps we should use that action more often. Nits are an example of this, I think. OTOH I'm not sure how easy github makes it to push to the PR branch. Addressing nits in-place should be at least as easy as making a nit-review-comment. Currently, I think there's a bit of friction in the UI because normally we start reading the diff, and if you want to make a change in-place, you have to scroll back to the top of the file, click "Edit", then find the spot in the now-open editor, make the change, then fill out a commit message and click commit to branch. Perhaps we should request an "edit-file-here" button accessible from the [+] button at the beginning of the line in the diff, which opens the diff comment UI. In fact, I will 🙂

![github-edit-in-place](https://user-images.githubusercontent.com/976081/44595108-1c81c600-a796-11e8-962c-c625b95ff50f.png)

A lot of the time, however, I think a review comment is either one where the reviewer would like to see the comment read by more people to spark discussion, or would like to describe a general feature desired of the modification, without actually re-writing the modification to include that feature. These are non-nits.

If we do end up pushing to the PR branch more often, we need to be careful where we draw the line, because AFAIK those changes will be attributed to the PRs author, so that puts us on a scale ranging from "oh, thanks, I totally didn't have time to address that review comment" to "that's totally not what I wanted to contribute with this PR" – although I'm sure the latter will be exceedingly unlikely because substantial changes need to be tested locally.

BridgeAR commented:
Any further progress here?

Trott commented:
I've added a 7-day "escape hatch" as suggested by @mcollina and endorsed by @BridgeAR and @mhdawson.

I've also simplified the language and grammar, removed passive voice, etc.

PTAL, especially @jasnell @ryzokuken @mcollina @mhdawson

tniessen commented:
Ping @mcollina @jasnell @ryzokuken

jasnell commented:
I still consider this change counter productive, completely unnecessary, and frivolous but won't block.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request-lite-pipeline/1004/

Trott commented:
Landed in b01e617
